env.sh

```diff
@@ -10,7 +10,7 @@
 # but you can still run scripts interactively from the commandline.
 # Also activates the virtualenv for this project.
 
-ENV_DIR=$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)
+ENV_DIR=$(cd $(dirname ${BASH_SOURCE[0]:-$0}) && pwd)
```

Comment:
Don't recognize this.  Can you elaborate?

Reply:
In zsh, **$0** refers to the "source" of the script, which in this case would be the string 'env.sh'.
When the script runs, the line `ENV_DIR=$(cd $(dirname ${BASH_SOURCE[0]:-$0}) && pwd)` is parsed by splitting it into its components and evaluating them one after another:
```
(1) ${BASH_SOURCE[0]:-$0}
(2) dirname ${BASH_SOURCE[0]:-$0}
(3) cd $(dirname ${BASH_SOURCE[0]:-$0})
(4) ENV_DIR=$(cd $(dirname ${BASH_SOURCE[0]:-$0}) && pwd)
```
Line 1 may be seen as a [variable substitution](https://www.tutorialspoint.com/unix/unix-shell-substitutions.htm) saying "use BASH_SOURCE[0] if it is defined; otherwise, use $0". I decided to add this "otherwise" condition when the original script didn't produce the result outlined in the script's comments; I'm running macOS Catalina (10.15.7), which uses _zsh_ as the default shell environment. 

Credit: https://stackoverflow.com/a/63072463